examples/non-fungible/src/contract.rs

```diff
@@ -0,0 +1,386 @@
+                    Self::check_account_authentication(
+                        None,
+                        runtime.authenticated_signer(),
+                        nft.owner,
+                    )?;
+
+                    self.transfer(nft, target_account).await
+                } else {
+                    self.remote_claim(source_account, token_id, target_account)
+                };
+
+                Ok(ApplicationCallOutcome {
+                    execution_outcome,
+                    ..Default::default()
+                })
+            }
+        }
+    }
+
+    async fn handle_session_call(
+        &mut self,
+        _runtime: &mut ContractRuntime,
+        _state: Self::SessionState,
+        _request: Self::SessionCall,
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<SessionCallOutcome<Self::Message, Self::Response, Self::SessionState>, Self::Error>
+    {
+        Err(Error::SessionsNotSupported)
+    }
+}
+
+impl NonFungibleToken {
+    /// Verifies that a transfer is authenticated for this local account.
+    fn check_account_authentication(
+        authenticated_application_id: Option<ApplicationId>,
+        authenticated_signer: Option<Owner>,
+        owner: AccountOwner,
+    ) -> Result<(), Error> {
+        match owner {
+            AccountOwner::User(address) if authenticated_signer == Some(address) => Ok(()),
+            AccountOwner::Application(id) if authenticated_application_id == Some(id) => Ok(()),
+            _ => Err(Error::IncorrectAuthentication),
+        }
+    }
+
+    /// Transfers the specified NFT to another account.
+    /// Authentication needs to have happened already.
+    async fn transfer(
+        &mut self,
+        mut nft: Nft,
+        target_account: Account,
+    ) -> ExecutionOutcome<Message> {
+        self.remove_nft(&nft).await;
+        if target_account.chain_id == system_api::current_chain_id() {
+            nft.owner = target_account.owner;
+            self.add_nft(nft).await;
+            ExecutionOutcome::default()
+        } else {
+            let message = Message::Transfer {
+                nft,
+                target_account,
+            };
+
+            ExecutionOutcome::default().with_tracked_message(target_account.chain_id, message)
+        }
+    }
+
+    async fn get_nft(&self, token_id: &TokenId) -> Nft {
+        self.nfts
+            .get(token_id)
+            .await
+            .expect("Failure in retrieving NFT")
+            .expect("NFT should not be None")
+    }
+
+    async fn mint(
+        &mut self,
+        owner: AccountOwner,
+        name: String,
+        payload: Vec<u8>,
+    ) -> ExecutionOutcome<Message> {
+        let token_id = Nft::create_token_id(
+            &system_api::current_chain_id(),
+            &system_api::current_application_id(),
+            &name,
+            &owner,
+            &payload,
+        );
+
+        self.add_nft(Nft {
+            token_id: token_id.clone(),
```

Comment:
Minor: This clone seems unnecessary

```suggestion
            token_id,
```